src/Week8/Party.hs

```diff
@@ -0,0 +1,185 @@
+ guest list for some subtree involves inviting that subtree’s boss, then
+ we are stuck, since we might want to consider inviting the boss of the
+ entire tree—in which case we don’t want to invite any of the subtree
+ bosses (since they wouldn’t have any fun anyway). But we might be
+ able to do better than just taking the best possible guest list for each
+ subtree and then excluding their bosses.
+
+ The solution is to generalize the recursion to compute more information, in
+ such a way that we can actually make the recursive step.
+
+ In particular, instead of just computing the best guest list for a given
+ tree, we will compute two guest lists:
+  1. the best possible guest list we can create if we invite the boss (that
+     is, the Employee at the root of the tree); and
+  2. the best possible guest list we can create if we don’t invite the boss.
+
+ It turns out that this gives us enough information at each step to
+ compute the optimal two guest lists for the next level up.
+-}
+
+----- Exercise 3 -----
+{-
+ Write a function
+   nextLevel :: Employee -> [(GuestList, GuestList)]  -> (GuestList, GuestList)
+
+ which takes two arguments. The first is the boss of the current subtree
+ (let's call him Bob). The second argument is a list of the results
+ for each subtree under Bob. Each result is a pair of GuestLists: the
+ first GuestList in the pair is the best possible guest list with the boss
+ of that subtree; the second is the best possible guest list without the
+ boss of that subtree. nextLevel should then compute the overall best
+ guest list that includes Bob, and the overall best guest list that doesn't
+ include Bob.
+-}
+nextLevel :: Employee -> [(GuestList, GuestList)] -> (GuestList, GuestList)
+nextLevel emp [] = (GL [emp] $ empFun emp, mempty)
+nextLevel emps gs = (withBoss, withoutBoss)
+  where
+    (withList, withoutList) = unzip gs
+    withBoss = glCons emps $ mconcat withoutList
+    withoutBoss = mconcat withList
+
+
+---- Exercise 4 -----
+{-
+ Finally, put all of this together to define
+     maxFun :: Tree Employee -> GuestList
+ which takes a company hierarchy as input and outputs a fun-maximizing
+ guest list. You can test your function on testCompany, provided in Employee.hs
+-}
+maxFun :: Tree Employee -> GuestList
+maxFun t = moreFun l r
+  where
+    (l, r) = foldTree nextLevel t
+
+
+---- Exercise 5 ------
+{-
+ Implement main :: IO () so that it reads your company’s hierarchy from the file
+ company.txt, and then prints out a formatted guest
+ list, sorted by first name, which looks like
+    Total fun: 23924
+    Adam Debergues
+    Adeline Anselme
+-}
+formatGuestList :: GuestList -> String
+formatGuestList (GL lst fun) = "Total fun: " ++ show fun ++ "\n" ++ unlines (empName <$> lst)
+
+main :: IO ()
+main = do
+  readFile "src/Week8/company.txt" >>= computeGuestList >>= putStr
+    where
+      computeGuestList = return . formatGuestList . maxFun . read
```

Comment:
Since you are using `do` notation, this can help with readability.
```suggestion
  input <- readFile "src/Week8/company.txt"
  let guestList = computeGuestList input
  putStr guestList
    where
      computeGuestList = formatGuestList . maxFun . read
```

Reply:
Done! Seems better than using the bind operator multiple times.